onchain/rollups/contracts/consensus/quorum/Quorum.sol

```diff
@@ -0,0 +1,133 @@
+
+    /// @notice The validator role.
+    /// @dev Only validators can submit claims.
+    bytes32 public constant VALIDATOR_ROLE = keccak256("VALIDATOR_ROLE");
+
+    /// @notice The history contract.
+    /// @dev See the `getHistory` function.
+    IHistory internal immutable history;
+
+    /// @notice For each claim, the set of validators that agree
+    /// that it should be submitted to the history contract.
+    mapping(bytes => EnumerableSet.AddressSet) internal yeas;
+
+    /// @notice An ERC-20 token transfer failed (`transfer` returned `false`).
+    error ERC20TokenTransferFailed();
+
+    /// @notice Construct a Quorum consensus
+    /// @param _validators the list of validators
+    constructor(IHistory _history, address[] memory _validators) {
+        // Iterate through the array of validators,
+        // and grant to each the validator role.
+        for (uint256 i; i < _validators.length; ++i) {
+            grantRole(VALIDATOR_ROLE, _validators[i]);
+        }
+
+        // Set history.
+        history = _history;
+    }
+
+    /// @notice Get the history contract.
+    /// @return The history contract
+    function getHistory() external view returns (IHistory) {
+        return history;
+    }
+
+    /// @notice Get a claim from the current history.
+    ///         The encoding of `_proofContext` might vary depending on the
+    ///         implementation of the current history contract.
+    /// @inheritdoc IConsensus
+    function getClaim(
+        address _dapp,
+        bytes calldata _proofContext
+    ) external view override returns (bytes32, uint256, uint256) {
+        return history.getClaim(_dapp, _proofContext);
+    }
+
+    /// @notice Submits a claim for voting.
+    ///         If this is the claim that reaches the majority, then
+    ///         the claim is submitted to the history contract.
+    ///         The encoding of `_claimData` might vary depending on the
+    ///         implementation of the current history contract.
+    /// @param _claimData Data for submitting a claim
+    /// @dev Can only be called by a validator,
+    ///      and the `Quorum` contract must have ownership over
+    ///      its current history contract.
+    function submitClaim(
+        bytes calldata _claimData
+    ) external onlyRole(VALIDATOR_ROLE) {
+        // Get the set of validators in favour of the claim
+        EnumerableSet.AddressSet storage claimYeas = yeas[_claimData];
+
+        // Add the message sender to such set.
+        claimYeas.add(msg.sender);
+
+        // Get number of validators in favour of the claim.
+        uint256 numOfVotesInFavour = claimYeas.length();
+
+        // Get the number of validators in the quorum.
+        uint256 quorumSize = getRoleMemberCount(VALIDATOR_ROLE);
+
+        // If this claim already has half of the quorum's approval,
+        // then we can submit it to the history contract.
+        if (numOfVotesInFavour > quorumSize / 2) {
```

Comment:
I think this should be a parameter instead of a majority.

Also, should not be ` x/2 + 1`? If there are three validators and we want the majority it results in a `1 of 3`.

Reply:
That is true. I misread as `>=`

Reply:
Actually, if we change to `==n` we solve the problem of including multiple times to the history. It would only add to History when the second validator (in this 2 of 3 example) post and ignore future validators.

Reply:
The current implementation of `History` prevents its owner from claiming the same data twice. :-)

Reply:
But I see what you mean. Maybe we should change this to `numOfVotesInFavour == 1 + quorumSize / 2`?